### PR TITLE
feat(tekton): adding command to create and apply metapipeline

### PR DESCRIPTION
**Pull request by hferentschik**
This is work in progress for the meta pipeline functionality. The pull request adds a new command `step create pipeline` which generates Tekton CRDs to run a meta pipeline allowing Apps to modify the source prior to the execution of the actual build pipeline.

This is work in progress but could be merged, since it does not affect the current behaviour. The next step is to let the meta pipeline call a modified `step create task`.

To run:
```bash
jx --verbose --log-level debug step create pipeline --clone-git-url <repo-url> --branch master --kind release --no-apply
```

In order to have some example apps, one can install any existing App and then edit its CRD adding:

```yaml
metadata:
  labels:
    jenkins.io/app-type: pipeline-extension
...
spec:
  pipelineExtension:
    args:
    - -l
    - /workspace/source
    command: ls
    image: alpine:latest
    name: My App
```

Relates to issue #3342

### CODE REVIEW

File: pkg/tekton/pipelines.go [hunk 1]
```diff
@@ -27,13 +27,17 @@ func GeneratePipelineActivity(buildNumber string, branch string, gitInfo *gits.G
 	name := gitInfo.Organisation + "-" + gitInfo.Name + "-" + branch + "-" + buildNumber
 	pipeline := gitInfo.Organisation + "/" + gitInfo.Name + "/" + branch
 	log.Infof("PipelineActivity for %s", name)
+	var toMerge map[string]string
```
Comment (hferentschik) [inline pkg/tekton/pipelines.go]: @abayer heads up, @dgozalo's latest changes related to batch builds changed the way you can use `step create task`.

Without this change here you cannot just call `jx step crate task` anymore. You also need to specify `--clone-git-url <repo>` and `-e PULL_REFS=<valid-pull-ref>`.

This change will get the old behaviour back.

File: pkg/jx/cmd/step/create/step_create_meta_pipeline.go [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+
+	err = o.handleResult(tektonClient, jxClient, tektonCRDs, buildNumber, o.Branch, o.PullRefs, ns, gitInfo)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *StepCreatePipelineOptions) validateCommandLineFlags() error {
```
Comment (hferentschik) [inline pkg/jx/cmd/step/create/step_create_meta_pipeline.go]: @abayer, I'd really like to end up at a stage where it is clearly defined what parameters are needed. Also,. easy validation should occur early on.

File: pkg/tekton/metapipeline/metapipeline.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+	}
+
+	// TODO - Add step which does call the 'step create task' functionality to build the actual build pipeline
+	return steps, nil
+}
+
+func buildEnvParams(params CRDCreationParameters) []corev1.EnvVar {
+	var envVars []corev1.EnvVar
+
+	envVars = append(envVars, corev1.EnvVar{
```
Comment (hferentschik) [inline pkg/tekton/metapipeline/metapipeline.go]: I might be missing some variables, but the ones here make obvious sense right now. If we need more and in particular if more is needed for creating the build pipeline needs to be seen.

File: pkg/tekton/metapipeline/metapipeline.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+		return nil, errors.Wrap(err, "error  retrieving pipeline contributor apps")
+	}
+	return appsList.Items, nil
+}
+
+// BuildExtensionSteps builds a step (container) for each of the apps specified.
+func BuildExtensionSteps(extendingApps []jenkinsv1.App) ([]syntax.Step, error) {
+	var steps []syntax.Step
+
+	// TODO - Add step to create effective pipeline. This way apps will always have a pipeline config to load and work with,
```
Comment (hferentschik) [inline pkg/tekton/metapipeline/metapipeline.go]: @abayer, so here I'd like to create the step which generates the effective jenkins-x.yml. Maybe extract that into a new `step syntax command`

File: pkg/tekton/metapipeline/metapipeline.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+			Image:     extension.Image,
+			Command:   extension.Command,
+			Arguments: extension.Args,
+		}
+
+		log.Debugf("App %s contributes with step %s", app.Name, util.PrettyPrint(step))
+		steps = append(steps, step)
+	}
+
+	// TODO - Add step which does call the 'step create task' functionality to build the actual build pipeline
```
Comment (hferentschik) [inline pkg/tekton/metapipeline/metapipeline.go]: @abayer, After the App steps, would come the step which calls the creation of the build pipeline CRDs. Effectively the current `step_create_task` minus the whole business of cloning and merging.

File: pkg/tekton/pipelines.go (hunk 1 above)
Comment (abayer) [reply pkg/tekton/pipelines.go]: Nice, good catch.

File: pkg/jx/cmd/step/create/step_create_meta_pipeline.go (hunk 2 above)
Comment (abayer) [reply pkg/jx/cmd/step/create/step_create_meta_pipeline.go]: Agreed.

File: pkg/tekton/metapipeline/metapipeline.go (hunk 4 above)
Comment (abayer) [reply pkg/tekton/metapipeline/metapipeline.go]: Yeah, I'm gonna do that in a separate PR today or tomorrow.

**Comment by hferentschik** [comment]
Closing this for now. I'll need to rebase and update the pull request.

**Pull request closed without merging**